A3A/addons/core/functions/FactoryOverhaul/fn_productionGUIControls.sqf

```diff
@@ -46,10 +86,31 @@ FO_fn_GUI_createFactoryTownPairs = {
 
 
 FO_fn_GUI_addWeaponsToListbox = {
+    params ["_category"];
+    lbClear IDC_FACTORY_WEPS;
+    if (_category isEqualTo "Any") then {
+        {
+            private _itemDisplayName  = getText (configFile >> "CfgWeapons" >> _x select 0 >> "displayName");
+            lbAdd [IDC_FACTORY_WEPS, _itemDisplayName];
+        } forEach call FO_fn_getUnlockedItems;    
+    } else {
+        {
+            private _itemDisplayName  = getText (configFile >> "CfgWeapons" >> _x select 0 >> "displayName");
+            private _itemCategory = _itemDisplayName call FO_fn_GUI_displayNameToCategory;
+            //private _wtf = call FO_fn_GUI_displayCategoryToItemCategory;
+            if (_itemCategory isEqualTo _category) then {
+                lbAdd [IDC_FACTORY_WEPS, _itemDisplayName];
+            };
+        } forEach call FO_fn_getUnlockedItems;    
+    };
+};
+
+
```

Comment:
FO_fn_GUI_addWeaponsToListbox is very messy, I don't think you need two loops. The code duplication of _itemDisplayName alone is enough to justify not doing that.

This should be refactored into one loop that assigns _itemDisplayName once and handles category checks. It's a bit slower, but significantly better to look at.

Reply:
Refactored.